[issues/106] - Added output/ hash to compiler cache file

filip-parity opened the pull request:
### TL;DR

This pull request addresses the **[caching issue](https://github.com/paritytech/foundry-polkadot/issues/106)** in existing Foundry’s compiler workflow by introducing an output directory hash to the compiler cache file. This ensures that changes to the output directory (such as those caused by switching compilers) are detected, and a full recompilation is triggered when necessary.

---

### Problem Statement

Previously, when running a sequence of builds with different compilers but the same output directory, Foundry’s cache would not detect that the output directory had been overwritten by a different compiler. This led to situations where a build was incorrectly skipped, even though the artifacts in the output directory were no longer valid for the current compiler.

For example, the test [must_rebuild_when_used_the_same_out](https://github.com/paritytech/foundry-polkadot/pull/105/files#diff-f62fc99e8363d8e17855f93dcbe4a42328420f669f11d5bdb4789cb905dbb7b5R15-R51) failed because:
- The first build with `solc` wrote artifacts to `resolc-out/` and updated the `solidity-files-cache.json`.
- The second build with `resolc` overwrote `resolc-out/` and updated a separate `resolc-solidity-files-cache.json`.
- The third build with `solc` again checked only the source file hashes and timestamps in `solidity-files-cache.json`, ignoring the fact that the output directory had been changed by `resolc`, and thus skipped recompilation.

This behavior, while consistent with the current design (and with forks like ZKSync), can lead to stale or incorrect artifacts when compilers are switched but the output directory is shared.

---

### Solution

- **Output Directory Hashing:**  
  - The cache now includes an `output_hash` field, which is a hash of all files in the output directory.
  - On each build, the current hash is compared to the cached hash. If they differ, the cache is invalidated and a full recompilation is triggered.
- **Implementation Details:**  
  - Added a method to recursively hash the contents of the output directory using MD5 and `walkdir`.
  - Updated the `CompilerCache` struct and logic to store and check this hash.
  - Updated dependencies to include `walkdir`.
- **Cache Invalidation:**  
  - If the output directory contents change (e.g., due to a different compiler run), the cache is cleared and a full rebuild is performed.

---

### Command Execution Breakdown

- **Command 1:** `forge build --out=resolc-out` (solc)  
  - Artifacts written to `resolc-out/`, cache updated.
- **Command 2:** `forge build --resolc` (resolc)  
  - Artifacts in `resolc-out/` overwritten, separate cache updated.
- **Command 3:** `forge build --out=resolc-out` (solc)  
  - **Previously**: Build skipped due to unchanged source files, even though output directory was changed.
  - **Now: Output directory hash mismatch triggers a full rebuild.**

---

### Analysis & Rationale

- This change ensures that the cache is always in sync with the actual build artifacts, preventing stale or incorrect outputs when compilers are switched or the output directory is externally modified.
- The solution is minimally invasive: it adds a single field to the cache and a lightweight hash computation, with no breaking changes to the public API.
- The approach is consistent with the current cache design and can be extended or adapted for other compilers or output directory scenarios.

---

### Additional Notes

- The new field is optional and will be populated on the next cache write, ensuring backward compatibility.
- This change does not affect the default behavior for users who do not switch compilers or share output directories.
- Dependency update: `walkdir` is now included in the workspace.

filip-parity commented:
Manual testing done _via_:
- Changes under `/paritytech/foundry-polkadot/Cargo.toml`:
```toml
foundry-compilers = { path = "../../foundry-compilers-polkadot/crates/compilers" }
```
and
```bash
cargo build # under `/paritytech/foundry-compilers-polkadot/`
```
- Running:
```bash
> rm -rf resolc-out/
> # rm -rf zkout/
/Users/filip/github/foundry-polkadot/target/debug/forge build --out=resolc-out
[⠰] Compiling...
Compiler run successful!
> # /Users/filip/github/foundry-polkadot/target/debug/forge build --out=zkout
> /Users/filip/github/foundry-polkadot/target/debug/forge build --resolc     # hardcoded to `resolc-out/`
[⠑] Compiling...
Compiler run successful with warnings:
Warning: Warning: Your code or one of its dependencies uses the 'extcodesize' instruction, which is
usually needed in the following cases:
  1. To detect whether an address belongs to a smart contract.
  2. To detect whether the deploy code execution has finished.
Polkadot comes with native account abstraction support (so smart contracts are just accounts
coverned by code), and you should avoid differentiating between contracts and non-contract
addresses.
--> lib/forge-std/src/StdCheats.sol
Warning: Warning: Your code or one of its dependencies uses the 'extcodesize' instruction, which is
usually needed in the following cases:
  1. To detect whether an address belongs to a smart contract.
  2. To detect whether the deploy code execution has finished.
Polkadot comes with native account abstraction support (so smart contracts are just accounts
coverned by code), and you should avoid differentiating between contracts and non-contract
addresses.
--> lib/forge-std/src/StdCheats.sol
Warning: Warning: Your code or one of its dependencies uses the 'extcodesize' instruction, which is
usually needed in the following cases:
  1. To detect whether an address belongs to a smart contract.
  2. To detect whether the deploy code execution has finished.
Polkadot comes with native account abstraction support (so smart contracts are just accounts
coverned by code), and you should avoid differentiating between contracts and non-contract
addresses.
--> lib/forge-std/src/StdUtils.sol
> # /Users/filip/github/foundry-polkadot/target/debug/forge build --zk-compile       # hardcoded to `zkout/`
> /Users/filip/github/foundry-polkadot/target/debug/forge build --out=resolc-out
[⠔] Compiling...
Compiler run successful!
> # /Users/filip/github/foundry-polkadot/target/debug/forge build --out=zkout
```

smiasojed commented on crates/compilers/src/cache.rs:
```diff
@@ -1055,6 +1062,27 @@ impl<'a, T: ArtifactOutput<CompilerContract = C::CompilerContract>, C: Compiler>
             // read the cache file if it already exists
             let mut cache = get_cache(project, invalidate_cache, preprocessed);
 
+            // Compare output directory hash with cached hash to detect changes.
```
Maybe this logic could be in `get_cache` function and if case of mismatch we would return `CompilerCache::new(Default::default(), paths, preprocessed)` WDYT?

filip-parity replied:
Yes, that's a good suggestion! Moving this logic to `get_cache` would make more sense because it's **more logical to handle all cache invalidation checks in one place** where we decide whether to use existing cache or create a new one and it actually **simplifies the flow** (if there's a hash mismatch, we just start with a fresh cache instead of clearing an existing one).

smiasojed commented on crates/compilers/src/cache.rs:
```diff
@@ -1243,6 +1271,9 @@ impl<'a, T: ArtifactOutput<CompilerContract = C::CompilerContract>, C: Compiler>
                 .strip_entries_prefix(project.root())
                 .strip_artifact_files_prefixes(project.artifacts_path());
+
+            cache.output_hash = Self::hash_directory(project.artifacts_path()).ok();
```
why we set output hash only if `write_to_disk` is set?

filip-parity replied:
Good catch, I set this to do some manual testing and I genuinely forgot to revert the change haha. **Addressing...**

smiasojed commented on crates/compilers/src/cache.rs:
```diff
@@ -1257,4 +1288,30 @@ impl<'a, T: ArtifactOutput<CompilerContract = C::CompilerContract>, C: Compiler>
     }
+
+    /// Hashes the contents of a directory.
+    pub fn hash_directory<P: AsRef<Path>>(directory: P) -> IoResult<String> {
```
does this function need to be public?

filip-parity replied:
Hmm not really, I made it `public` just in case we would need to hash other directories as part of the compiler caching but I guess we can disregard this until (if) the need arise. **Addressing...**

smiasojed commented:
Thank you. Overall, it looks good.

filip-parity commented:
> dq: can you give steps to reproduce the line here?
> 
> > The second build with resolc overwrote resolc-out/ and updated a separate resolc-solidity-files-cache.json.
> 
> when i recompile either with resolc or solc - only `solidity-files-cache.json` is being updated.

**FYI** I added a more comprehensive analysis as part of [this issue](https://github.com/paritytech/foundry-polkadot/issues/106) but TL;DR commands are:
```bash
$(FORGE) build --out=resolc-out
$(FORGE) build --resolc	# hardcoded to `resolc-out/`
$(FORGE) build --out=resolc-out
```

Also, you have the initial test within another pull request that reproduced this issue available [here](https://github.com/paritytech/foundry-polkadot/pull/105/files#diff-f62fc99e8363d8e17855f93dcbe4a42328420f669f11d5bdb4789cb905dbb7b5R15-R51).

filip-parity commented:
> > > dq: can you give steps to reproduce the line here?
> > > > The second build with resolc overwrote resolc-out/ and updated a separate resolc-solidity-files-cache.json.
> > > 
> > > 
> > > when i recompile either with resolc or solc - only `solidity-files-cache.json` is being updated.
> > 
> > 
> > **FYI** I added a more comprehensive analysis as part of [this issue](https://github.com/paritytech/foundry-polkadot/issues/106) but TL;DR commands are:
> > ```shell
> > $(FORGE) build --out=resolc-out
> > $(FORGE) build --resolc	# hardcoded to `resolc-out/`
> > $(FORGE) build --out=resolc-out
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Also, you have the initial test within another pull request that reproduced this issue available [here](https://github.com/paritytech/foundry-polkadot/pull/105/files#diff-f62fc99e8363d8e17855f93dcbe4a42328420f669f11d5bdb4789cb905dbb7b5R15-R51).
> 
> oof, accidentaly used wrong flag for resolc
> 
> why not do something like this? it should also invalidate caches It will remove unrelated build-infos from the out folder
> 
> ```
> diff --git a/crates/compilers/src/buildinfo.rs b/crates/compilers/src/buildinfo.rs
> ...
> +}
> ```

Hmm interesting, LGTM but my Rust knowledge is still in kindergarten. The proposed changes have even less footprint on the existing codebase. **Any opinion here @smiasojed ?**

filip-parity commented:
> We need to find a clean solution if you thing that this is better, then please use it.

The build info approach actually provides MORE information because:
- It captures both input and output state
- It maintains direct linkage between source files and their artifacts
- It includes compilation settings and dependencies
- The cleanup of orphaned build files ensures consistency

The new approach is actually more comprehensive and precise than just hashing the output directory. I updated the current implementation. Thank you @pkhry !
